pacwrapper.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"bytes"
+	"log"
+	"net/http"
+	"text/template"
+)
+
+type PACData struct {
+	Port     int
+	PACURL   string
+	Domain   string
+	Username string
+}
+
+type pacData struct {
+	PACData
+	PAC string
+}
+
+type PACWrapper struct {
+	data      pacData
+	tmpl      *template.Template
+	alpacaPAC string
+}
+
+// PACWrapper template for serving a PAC file to point at alpaca or DIRECT.
+// If we have a valid PAC file, we wrap that PAC file with a wrapper
+// function that only returns "DIRECT" or "localhost:port". If we do not
+// have a PAC file, the PAC function we serve only returns "DIRECT", which
+// should prevent all requests reaching us.
+var pacWrapTmpl = `// Wrapped for and by alpaca
+function FindProxyForURL(url, host) {
+{{ if .PAC }}
+  return FindProxyForURL(url, host) === "DIRECT" ? "DIRECT" : "PROXY localhost:{{.Port}}";
+{{.PAC}}
+{{ else }}
+  return "DIRECT";
+{{ end }}
+}
+`
+
+func NewPACWrapper(data PACData) *PACWrapper {
+	t := template.Must(template.New("alpaca").Parse(pacWrapTmpl))
+	return &PACWrapper{pacData{data, ""}, t, ""}
+}
+
+func (pw *PACWrapper) Wrap(pacjs []byte) {
+	pac := string(pacjs)
+	if pac == pw.data.PAC && pw.alpacaPAC != "" {
+		return
+	}
+	pw.data.PAC = pac
+	b := &bytes.Buffer{}
+	if err := pw.tmpl.Execute(b, pw.data); err != nil {
+		log.Printf("error executing PAC wrap template: %v\n", err)
+		return
+	}
+	pw.alpacaPAC = b.String()
+}
+
+func (pw *PACWrapper) SetupHandlers(mux *http.ServeMux) {
+	mux.HandleFunc("/proxy.pac", pw.handleProxyPAC)
+	mux.HandleFunc("/alpaca.pac", pw.handleAlpacaPAC)
+}
+
+func (pw *PACWrapper) handleProxyPAC(w http.ResponseWriter, req *http.Request) {
+	pw.handlePAC(w, req, pw.data.PAC)
+}
+
+func (pw *PACWrapper) handleAlpacaPAC(w http.ResponseWriter, req *http.Request) {
+	pw.handlePAC(w, req, pw.alpacaPAC)
+}
```

Comment:
nit: I'd be sorely tempted to inline these in `SetupHandlers`.

Reply:
I was tempted and I also tried another approach of type aliasing `string` to `pacString` and defining a `ServeHTTP` method on it, so as to remove these two single line functions. But the anonymous function I felt was a little too verbose in the function signature (`w http.ResponseWriter, req *http.Request)` is a little to long with the func and embedded in the function call), and the type alias didn't work unless it was a pointer which had an impact on `text/template` use of that.

This does feel like a little bit of a wart but that seems to happen somewhat often with Go. It is quite readable as is, so I'll stick with it unless there's something I find definitively better.